File: arrays_hashing/top_k_frequent_elements/top_k_freq_elements_test.go
package topkfrequent

import (
	"reflect"
	"sort"
	"testing"
)

func TestTopKFrequent(t *testing.T) {
	tests := []struct {
		name     string
		nums     []string
		numsInt  []int
		k        int
		expected []int
	}{
		{
			name:     "Example 1",
			numsInt:  []int{1, 2, 2, 3, 3, 3},
			k:        2,
			expected: []int{2, 3},
		},
		{
			name:     "Example 2",
			numsInt:  []int{7, 7},
			k:        1,
			expected: []int{7},
		},
		{
			name:     "Single Element",
			numsInt:  []int{1},
			k:        1,
			expected: []int{1},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			actual := topKFrequent(tt.numsInt, tt.k)

			// Since order doesn't matter, sort both before comparing
			sort.Ints(actual)
			sort.Ints(tt.expected)
			if !reflect.DeepEqual(actual, tt.expected) {
				t.Errorf("topKFrequent() = %v, want %v", actual, tt.expected)
			}
		})
	}
}

File: arrays_hashing/top_k_frequent_elements/top_k_freq_elements_test.py
import pytest
from top_k_freq_elements import top_k_freq_elements


@pytest.mark.parametrize(
    "nums, k, expected",
    [
        ([1, 1, 1, 2, 2, 3], 2, [1, 2]),
        ([7, 7], 1, [7]),
        ([1], 1, [1]),
        ([-1, -1], 1, [-1]),
        ([4, 1, -1, 2, -1, 2, 3], 2, [-1, 2]),
        ([1, 2], 2, [1, 2]),
    ],
)
def test_top_k_frequent(nums, k, expected):
    # Call the function directly
    actual = top_k_freq_elements(nums, k)

    # Using set comparison because the output order doesn't matter
    assert set(actual) == set(expected), (
        f"Failed on {nums}. Expected {expected}, got {actual}"
    )


def test_k_equals_length():
    nums = [1, 2, 3]
    k = 3
    actual = top_k_freq_elements(nums, k)
    assert set(actual) == {1, 2, 3}
